smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/DirectedTypeScriptCodegen.java

```diff
@@ -0,0 +1,480 @@
+        directive.context().writerDelegator().useShapeWriter(directive.shape(), writer -> {
+            UnionGenerator generator = new UnionGenerator(
+                    directive.model(),
+                    directive.symbolProvider(),
+                    writer,
+                    directive.shape(),
+                    directive.settings().generateServerSdk()
+            );
+            generator.run();
+        });
+    }
+
+    @Override
+    public void generateEnumShape(GenerateEnumDirective<TypeScriptCodegenContext, TypeScriptSettings> directive) {
+        directive.context().writerDelegator().useShapeWriter(directive.shape(), writer -> {
+            EnumGenerator generator = new EnumGenerator(
+                    directive.shape().asStringShape().get(),
+                    directive.symbolProvider().toSymbol(directive.shape()),
+                    writer
+            );
+            generator.run();
+        });
+    }
+
+    @Override
+    public void customizeBeforeIntegrations(
+            CustomizeDirective<TypeScriptCodegenContext, TypeScriptSettings> directive) {
+        // Write shared / static content.
+        STATIC_FILE_COPIES.forEach((from, to) -> {
+            LOGGER.fine(() -> "Writing contents of `" + from + "` to `" + to + "`");
+            directive.fileManifest().writeFile(from, getClass(), to);
+        });
+
+        // TODO: do all of these parts below are before/after?
+        SymbolVisitor.writeModelIndex(directive.connectedShapes().values(), directive.symbolProvider(),
+                directive.fileManifest());
+
+        // Generate the client Node and Browser configuration files. These
+        // files are switched between in package.json based on the targeted
+        // environment.
+        if (directive.settings().generateClient()) {
+            // For now these are only generated for clients.
+            // TODO: generate ssdk config
+            RuntimeConfigGenerator configGenerator = new RuntimeConfigGenerator(
+                    directive.settings(),
+                    directive.model(),
+                    directive.symbolProvider(),
+                    directive.context().writerDelegator(),
+                    directive.context().integrations());
+            for (LanguageTarget target : LanguageTarget.values()) {
+                LOGGER.fine("Generating " + target + " runtime configuration");
+                configGenerator.generate(target);
+            }
+        }
+
+        // Write each custom file.
+        for (TypeScriptIntegration integration : directive.context().integrations()) {
+            LOGGER.finer(() -> "Calling writeAdditionalFiles on " + integration.getClass().getCanonicalName());
+            integration.writeAdditionalFiles(
+                    directive.settings(),
+                    directive.model(),
+                    directive.symbolProvider(),
+                    directive.context().writerDelegator()::useFileWriter);
+        }
```

Comment:
I think this loop here can be removed by moving `integration.writeAdditionalFiles` to `integration.customize` which CodegenDirector will call.

Reply:
It will be awesome if we can do it in the same customization loop.

Reply:
Yeah, I think this can be done as a refactoring later.